--- a/src/main/java/io/papermc/paper/util/Holderable.java
+++ b/src/main/java/io/papermc/paper/util/Holderable.java
@@ -33,7 +_,7 @@
                 }
                 yield registry.get(Key.key(string));
             }
-            case JsonObjectWrapper(final JsonObject element) -> {
+            case JsonObjectWrapper(JsonObject element) -> {
                 if (!(registry instanceof final CraftRegistry<?, ?> craftRegistry) || !craftRegistry.supportsDirectHolders()) {
                     throw new IllegalArgumentException("Cannot deserialize direct holders for " + registry);
                 }
